src/Raven.Server/Documents/ShardedHandlers/ShardedQueriesHandler.cs

```diff
@@ -0,0 +1,843 @@
+                }
+                else
+                {
+                    _filteredShardIndexes = null;
+                }
+            }
+
+            public long ResultsEtag => _result.ResultEtag;
+
+            public void Initialize()
+            {
+                if (_query.Metadata.IsGraph)
+                {
+                    // * Graph queries - not supported for sharding
+                    throw new NotSupportedException("Graph queries aren't supported for sharding");
+                }
+
+                // now we have the index query, we need to process that and decide how to handle this.
+                // There are a few different modes to handle:
+                var queryTemplate = _query.ToJson(_context);
+                if (_query.Metadata.IsCollectionQuery)
+                {
+                    // * For collection queries that specify ids, we can turn that into a set of loads that 
+                    //   will hit the known servers
+
+                    (List<string> ids, string _) = _query.ExtractIdsFromQuery(_parent.ServerStore, _parent.ShardedContext.DatabaseName);
+                    if (ids != null)
+                    {
+                        GenerateLoadByIdQueries(ids, _commands);
+                        return;
+                    }
+                }
+                
+                // * For paging queries, we modify the limits on the query to include all the results from all
+                //   shards if there is an offset. But if there isn't an offset, we can just get the limit from
+                //   each node and then merge them
+                RewriteQueryForPaging(ref queryTemplate);
+
+                // * If we have a projection in a map-reduce index,
+                //   the shards will send the query result and the orchestrator will re-reduce and apply the projection
+                //   in that case we must send the query without the projection
+                RewriteQueryForProjection(ref queryTemplate);
+
+                // * For collection queries that specify startsWith by id(), we need to send to all shards
+                // * For collection queries without any where clause, we need to send to all shards
+                // * For indexes, we sent to all shards
+                for (int i = 0; i < _parent.ShardedContext.NumberOfShardNodes; i++)
+                {
+                    if (_filteredShardIndexes?.Contains(i) == false)
+                        continue;
+
+                    _commands[i] = new ShardedQueryCommand(_parent, queryTemplate, _query.Metadata.IndexName);
+                }
+            }
+
+            private void RewriteQueryForPaging(ref BlittableJsonReaderObject queryTemplate)
+            {
+                if (_query.Offset is null or <= 0)
+                    return;
+
+                var clone = _query.Metadata.Query.ShallowCopy();
+                clone.Offset = null; // sharded queries has to start from 0 on all nodes
```

Comment:
this is very problematic for deep paging... we will get the entire result set.
maybe we need to prefer performance over accuracy for deep paging.

Reply:
any comments?

Reply:
https://issues.hibernatingrhinos.com/issue/RavenDB-13109